hs_access_control/tests/test_communities.py

```diff
@@ -0,0 +1,646 @@
+        ggp = UserCommunityProvenance.objects.get(user=self.dog2, community=self.pets)
+        self.assertEqual(ggp.privilege, PrivilegeCodes.VIEW)
+        ggp = UserCommunityProvenance.objects.get(user=self.cat2, community=self.pets)
+        self.assertEqual(ggp.privilege, PrivilegeCodes.VIEW)
+
+        self.assertEqual(self.pets.get_effective_user_privilege(self.dog2),
+                         PrivilegeCodes.VIEW)
+        self.assertEqual(self.pets.get_effective_user_privilege(self.cat2),
+                         PrivilegeCodes.VIEW)
+
+        self.assertTrue(self.cat2 in self.pets.member_users)
+        self.assertTrue(self.dog2 in self.pets.member_users)
+
+        # accept ownership of community by a user
+        self.assertTrue(self.dog.uaccess.can_share_community_with_user(self.pets, self.dog2,
+                                                                        PrivilegeCodes.OWNER))
+        self.dog.uaccess.share_community_with_user(self.pets, self.dog2,
+                                                   PrivilegeCodes.OWNER)
+        # privilege object created
+        ggp = UserCommunityPrivilege.objects.get(user=self.dog2, community=self.pets)
+        self.assertEqual(ggp.privilege, PrivilegeCodes.OWNER)
+
+        # provenance object created
+        ggp = UserCommunityProvenance.objects.filter(user=self.dog2, community=self.pets)
+        self.assertEqual(ggp.count(), 2)
+
+        # Privileges are changed by the previous act
+        self.assertEqual(self.pets.get_effective_user_privilege(self.dog2),
+                         PrivilegeCodes.OWNER)
+        self.assertEqual(self.pets.get_effective_user_privilege(self.cat2),
+                         PrivilegeCodes.VIEW)
+
+        # downgrade share privilege
+        self.dog.uaccess.share_community_with_user(self.pets, self.dog2, PrivilegeCodes.CHANGE)
+
+        # privilege object created
+        ggp = UserCommunityPrivilege.objects.get(user=self.dog2, community=self.pets)
+        self.assertEqual(ggp.privilege, PrivilegeCodes.CHANGE)
+
+        # provenance object created
+        ggp = UserCommunityProvenance.objects.filter(user=self.dog2, community=self.pets)
+        self.assertEqual(ggp.count(), 3)
+
+        self.assertEqual(self.pets.get_effective_user_privilege(self.dog2),
+                         PrivilegeCodes.CHANGE)
+        self.assertEqual(self.pets.get_effective_user_privilege(self.cat2),
+                         PrivilegeCodes.VIEW)
+
+        # unshare community with user
+        self.assertTrue(self.dog.uaccess.can_unshare_community_with_user(self.pets, self.dog2))
+        self.dog.uaccess.unshare_community_with_user(self.pets, self.dog2)
+
+        self.assertEqual(self.pets.get_effective_user_privilege(self.dog2),
+                         PrivilegeCodes.NONE)
+        self.assertEqual(self.pets.get_effective_user_privilege(self.cat2),
+                         PrivilegeCodes.VIEW)
+
```

Comment:
May also test here who are the users of the community(pets) after unsharing user 'dog2' with the community.

Reply:
Later.